### ISSUE TITLE
Linux SSL library documentation

### ISSUE THREAD

**Issue by Rudis1261** [brianmario/mysql2#1246]
I wonder if someone is able to explain what installing the required libs and compiling the gem to support SSL connections?

I have tried on both `ubuntu` and `debian` but I have not been able to. I ensured that:
```shell
apt-get install -y ruby-dev build-essential
apt-get install -y openssl libmariadb-dev
```

The MySQL client is able to connect to the host on SSL:
```shell
→ mysql -h <host> --ssl -uruditest -p

MariaDB [(none)]> status
--------------
SSL:			Cipher in use is ECDHE-RSA-AES256-GCM-SHA384
```

But when attempting to use SSL when using the mysql2 gem I get:
```shell
/usr/local/bundle/gems/mysql2-0.5.3/lib/mysql2/client.rb:51: warning: Your mysql client library does not support ssl_mode as expected.
```

It seems that modern openssl dev libs now live here:
`/usr/include/openssl`

I have also tried:
`gem install mysql2 -- --with-opt-lib=/usr/include/openssl`

And:
`gem install mysql2 -- --with-cppflags=-I/usr/include/openssl --with-ldflags=-L/usr/include/openssl`

The results of `/usr/include/openssl`:
```shell
root@206d5396fbe3:/# ls -alhtr /usr/include/openssl
total 1.6M
-rw-r--r-- 1 root root 8.6K Oct 12  2019 x509v3err.h
-rw-r--r-- 1 root root  33K Oct 12  2019 x509v3.h
-rw-r--r-- 1 root root 6.7K Oct 12  2019 x509err.h
-rw-r--r-- 1 root root  32K Oct 12  2019 x509_vfy.h
-rw-r--r-- 1 root root  43K Oct 12  2019 x509.h
-rw-r--r-- 1 root root 1.4K Oct 12  2019 whrlpool.h
-rw-r--r-- 1 root root 2.7K Oct 12  2019 uierr.h
-rw-r--r-- 1 root root  16K Oct 12  2019 ui.h
-rw-r--r-- 1 root root 1.7K Oct 12  2019 txt_db.h
-rw-r--r-- 1 root root 6.6K Oct 12  2019 tserr.h
-rw-r--r-- 1 root root  22K Oct 12  2019 ts.h
-rw-r--r-- 1 root root  71K Oct 12  2019 tls1.h
-rw-r--r-- 1 root root 1.3K Oct 12  2019 symhacks.h
-rw-r--r-- 1 root root 4.3K Oct 12  2019 storeerr.h
-rw-r--r-- 1 root root  11K Oct 12  2019 store.h
-rw-r--r-- 1 root root 3.1K Oct 12  2019 stack.h
-rw-r--r-- 1 root root  46K Oct 12  2019 sslerr.h
-rw-r--r-- 1 root root  15K Oct 12  2019 ssl3.h
-rw-r--r-- 1 root root  542 Oct 12  2019 ssl2.h
-rw-r--r-- 1 root root 109K Oct 12  2019 ssl.h
-rw-r--r-- 1 root root 1.3K Oct 12  2019 srtp.h
-rw-r--r-- 1 root root 3.8K Oct 12  2019 srp.h
-rw-r--r-- 1 root root 3.8K Oct 12  2019 sha.h
-rw-r--r-- 1 root root 3.4K Oct 12  2019 seed.h
-rw-r--r-- 1 root root 8.0K Oct 12  2019 safestack.h
-rw-r--r-- 1 root root 8.9K Oct 12  2019 rsaerr.h
-rw-r--r-- 1 root root  22K Oct 12  2019 rsa.h
-rw-r--r-- 1 root root 1.3K Oct 12  2019 ripemd.h
-rw-r--r-- 1 root root 2.0K Oct 12  2019 rc5.h
-rw-r--r-- 1 root root  825 Oct 12  2019 rc4.h
-rw-r--r-- 1 root root 1.5K Oct 12  2019 rc2.h
-rw-r--r-- 1 root root 4.5K Oct 12  2019 randerr.h
-rw-r--r-- 1 root root 4.7K Oct 12  2019 rand_drbg.h
-rw-r--r-- 1 root root 2.2K Oct 12  2019 rand.h
-rw-r--r-- 1 root root 5.0K Oct 12  2019 pkcs7err.h
-rw-r--r-- 1 root root  12K Oct 12  2019 pkcs7.h
-rw-r--r-- 1 root root 3.7K Oct 12  2019 pkcs12err.h
-rw-r--r-- 1 root root 9.7K Oct 12  2019 pkcs12.h
-rw-r--r-- 1 root root 5.0K Oct 12  2019 pemerr.h
-rw-r--r-- 1 root root  415 Oct 12  2019 pem2.h
-rw-r--r-- 1 root root  16K Oct 12  2019 pem.h
-rw-r--r-- 1 root root 6.1K Oct 12  2019 ossl_typ.h
-rw-r--r-- 1 root root 4.1K Oct 12  2019 opensslv.h
-rw-r--r-- 1 root root 3.3K Oct 12  2019 ocsperr.h
-rw-r--r-- 1 root root  15K Oct 12  2019 ocsp.h
-rw-r--r-- 1 root root 1.3K Oct 12  2019 objectserr.h
-rw-r--r-- 1 root root 6.5K Oct 12  2019 objects.h
-rw-r--r-- 1 root root 213K Oct 12  2019 obj_mac.h
-rw-r--r-- 1 root root  11K Oct 12  2019 modes.h
-rw-r--r-- 1 root root 1.1K Oct 12  2019 mdc2.h
-rw-r--r-- 1 root root 1.3K Oct 12  2019 md5.h
-rw-r--r-- 1 root root 1.3K Oct 12  2019 md4.h
-rw-r--r-- 1 root root 1.1K Oct 12  2019 md2.h
-rw-r--r-- 1 root root 9.1K Oct 12  2019 lhash.h
-rw-r--r-- 1 root root 2.1K Oct 12  2019 kdferr.h
-rw-r--r-- 1 root root 4.3K Oct 12  2019 kdf.h
-rw-r--r-- 1 root root 2.1K Oct 12  2019 idea.h
-rw-r--r-- 1 root root 1.6K Oct 12  2019 hmac.h
-rw-r--r-- 1 root root  12K Oct 12  2019 evperr.h
-rw-r--r-- 1 root root  74K Oct 12  2019 evp.h
-rw-r--r-- 1 root root  12K Oct 12  2019 err.h
-rw-r--r-- 1 root root 5.4K Oct 12  2019 engineerr.h
-rw-r--r-- 1 root root  34K Oct 12  2019 engine.h
-rw-r--r-- 1 root root  16K Oct 12  2019 ecerr.h
-rw-r--r-- 1 root root  358 Oct 12  2019 ecdsa.h
-rw-r--r-- 1 root root  358 Oct 12  2019 ecdh.h
-rw-r--r-- 1 root root  63K Oct 12  2019 ec.h
-rw-r--r-- 1 root root  924 Oct 12  2019 ebcdic.h
-rw-r--r-- 1 root root 8.7K Oct 12  2019 e_os2.h
-rw-r--r-- 1 root root 1.6K Oct 12  2019 dtls1.h
-rw-r--r-- 1 root root 3.0K Oct 12  2019 dsaerr.h
-rw-r--r-- 1 root root 9.5K Oct 12  2019 dsa.h
-rw-r--r-- 1 root root 3.9K Oct 12  2019 dherr.h
-rw-r--r-- 1 root root  14K Oct 12  2019 dh.h
-rw-r--r-- 1 root root 7.5K Oct 12  2019 des.h
-rw-r--r-- 1 root root 3.4K Oct 12  2019 cterr.h
-rw-r--r-- 1 root root  16K Oct 12  2019 ct.h
-rw-r--r-- 1 root root 2.3K Oct 12  2019 cryptoerr.h
-rw-r--r-- 1 root root  17K Oct 12  2019 crypto.h
-rw-r--r-- 1 root root 3.4K Oct 12  2019 conferr.h
-rw-r--r-- 1 root root 1.3K Oct 12  2019 conf_api.h
-rw-r--r-- 1 root root 5.5K Oct 12  2019 conf.h
-rw-r--r-- 1 root root 1.2K Oct 12  2019 comperr.h
-rw-r--r-- 1 root root 1.3K Oct 12  2019 comp.h
-rw-r--r-- 1 root root  11K Oct 12  2019 cmserr.h
-rw-r--r-- 1 root root  16K Oct 12  2019 cms.h
-rw-r--r-- 1 root root 1.1K Oct 12  2019 cmac.h
-rw-r--r-- 1 root root 1.7K Oct 12  2019 cast.h
-rw-r--r-- 1 root root 3.2K Oct 12  2019 camellia.h
-rw-r--r-- 1 root root  820 Oct 12  2019 buffererr.h
-rw-r--r-- 1 root root 1.6K Oct 12  2019 buffer.h
-rw-r--r-- 1 root root 4.8K Oct 12  2019 bnerr.h
-rw-r--r-- 1 root root  22K Oct 12  2019 bn.h
-rw-r--r-- 1 root root 1.9K Oct 12  2019 blowfish.h
-rw-r--r-- 1 root root 6.3K Oct 12  2019 bioerr.h
-rw-r--r-- 1 root root  35K Oct 12  2019 bio.h
-rw-r--r-- 1 root root 1.3K Oct 12  2019 asyncerr.h
-rw-r--r-- 1 root root 2.4K Oct 12  2019 async.h
-rw-r--r-- 1 root root  33K Oct 12  2019 asn1t.h
-rw-r--r-- 1 root root  15K Oct 12  2019 asn1err.h
-rw-r--r-- 1 root root  395 Oct 12  2019 asn1_mac.h
-rw-r--r-- 1 root root  33K Oct 12  2019 asn1.h
-rw-r--r-- 1 root root 3.3K Oct 12  2019 aes.h
drwxr-xr-x 2 root root 4.0K Mar 31  2020 .
drwxr-xr-x 1 root root 4.0K Feb 28 10:16 ..
```

**Comment by trbroyles1** [comment]
+1, I have the exact same issue as described above

**Comment by trbroyles1** [comment]
@Rudis1261 having just spent the last 3 hours trying to get this working I know the pain you are feeling.

With that in mind, I thought I'd add a note on what finally worked for me.

It looks like this is -- from browsing the other reported issues here -- yet another case of the currently-released gem 0.5.3 being out of sync with the current version of the MySQL / MariaDB clients.

In the end I pulled the git version of the gem and used it, and that got things working.

So, in my Gemfile for my project (I'm using bundler), I replaced...
`gem "mysql2", "~> 0.5.3"`
with...
`gem "mysql2", git: "https://github.com/brianmario/mysql2"`
and then did a `bundle install` to update everything.

after that, I was able to get mysql2 and active record to connect using SSL with a URL-formatted connection string, like so:
`mysql2://user:pass@host:3306/dbname?ssl_mode=REQUIRED`

Sad that we have to go to github for the current version of the gem -- but hey at least it is a working solution.

Hopefully this helps you make some progress !

**Comment by Rudis1261** [comment]
Shot thanks man, lemme try this out and report back :sweat_smile:

**Comment by Rudis1261** [comment]
You legend! Thanks so much!

**Comment by Rudis1261** [comment]
Reopening issue. For us as an org it is important to be able to pin to a specific version. To prevent us from suffering failures as a result of ruby version upgrades. We are typically a ruby version or two behind. 

Not sure what the answer is, but it's something that probably needs addressing, as searching for info on the topic. I saw a few issues relating to this.

**Comment by sodabrew** [comment]
Included in the 0.5.4 release! install from git should no longer be required for this issue.
